src/features/authentication/public/hooks/useLoginForm.ts

```diff
@@ -0,0 +1,82 @@
+import { useMemo, useState } from 'react';
+import { SubmitHandler, useForm } from 'react-hook-form';
+import { useNavigate } from 'react-router';
+import Result, { isOk } from 'true-myth/result';
+
+import useAuth from '~/features/authentication/public/hooks/useAuth';
+
+interface LoginForm {
+    password: string;
+    username: string;
+}
+
+export default function useLoginForm() {
+    const { authenticate } = useAuth();
+    const navigate = useNavigate();
+    const [submitErrors, setSubmitErrors] = useState<null | Result<boolean, { reason: string }>>(null);
+
+    const {
+        formState: { dirtyFields, errors, isDirty, isSubmitted, isValid },
+        handleSubmit,
+        register,
+    } = useForm<LoginForm>({
+        defaultValues: { password: '', username: '' },
+    });
```

Comment:
was able to simplify the login form logic quite a bit, not having to use react-hook-form's manual `watch` etc APIs, and I think i maintained the current behavior that we want with that.